espresso/environment/9_pipeline_enhancement_test.go

```diff
@@ -0,0 +1,98 @@
+//		Arrange:
+//			Running Sequencer, Batcher in Espresso mode
+//		Act:
+//			Send a transaction not signed by the batcher (which means it will revert) to the inbox contract with a specific sequence of bytes e.g. 0x42424242424242424242
+//          Fetch N, the block number where the transaction was included (even though it is reverted)
+//		Assert:
+//			Instantiate a CalldataSource object with block N. CalldataSource is the object that reads the calldata from the inbox contract and filters it using the isValidBatchTx function.
+//			Verify that the concatenated bytes of all batch transactions does not include 0x42424242424242424242
+
+func TestPipelineEnhancement(t *testing.T) {
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	launcher := new(env.EspressoDevNodeLauncherDocker)
+
+	system, espressoDevNode, err := launcher.StartDevNet(ctx, t)
+	// Signal the testnet to shut down
+	if have, want := err, error(nil); have != want {
+		t.Fatalf("failed to start dev environment with espresso dev node:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+	}
+
+	l1Client := system.NodeClient(e2esys.RoleL1)
+
+	defer env.Stop(t, system)
+	defer env.Stop(t, espressoDevNode)
+
+	// Send a transaction not signed by the batcher to the inbox contract
+	// Create the transaction
+
+	txData := []byte("42424242424242424242")
+
+	tx := gethTypes.MustSignNewTx(system.Cfg.Secrets.Bob, system.RollupConfig.L1Signer(), &gethTypes.DynamicFeeTx{
+		ChainID:   system.Cfg.L1ChainIDBig(),
+		Nonce:     0,
+		GasTipCap: big.NewInt(1 * params.GWei),
+		GasFeeCap: big.NewInt(10 * params.GWei),
+		Gas:       5_000_000,
+		To:        &system.RollupConfig.BatchInboxAddress,
+		Value:     big.NewInt(0),
+		Data:      txData,
+	})
+
+	l := log.NewLogger(slog.Default().Handler())
+	err = l1Client.SendTransaction(ctx, tx)
+	require.NoError(t, err)
+
+	receipt, err := wait.ForReceiptFail(ctx, l1Client, tx.Hash())
+	require.NoError(t, err, "Waiting for receipt on transaction", tx)
+
+	l1ClientFetching, _ := client.NewRPC(ctx, nil, system.NodeEndpoint(e2esys.RoleL1).RPC())
+	l1RefClient, err := sources.NewL1Client(l1ClientFetching, l, nil, sources.L1ClientDefaultConfig(system.RollupConfig, true, sources.RPCKindStandard))
+
+	system.RollupConfig.EcotoneTime = nil
+	factory := derive.NewDataSourceFactory(l, system.RollupConfig, l1RefClient, nil, nil)
+
+	batcherAddress := crypto.PubkeyToAddress(*system.BatchSubmitter.BatcherPublicKey)
+	l1Block, err := l1Client.BlockByNumber(ctx, receipt.BlockNumber)
+	require.NoError(t, err)
+	l1BlockRef := eth.L1BlockRef{
+		Hash:       l1Block.Hash(),
+		Number:     l1Block.NumberU64(),
+		ParentHash: l1Block.ParentHash(),
+		Time:       l1Block.Time(),
+	}
+	datas, err := factory.OpenData(ctx, l1BlockRef, batcherAddress)
+	require.NoError(t, err)
+
+	data, err := datas.Next(ctx)
+
+	// The L1 data collected by the derivation pipeline is empty because the batch information has been discarded
```

Comment:
Nit: I think we should disable the batcher for this test then, otherwise if it's fast enough (probably not something we'll hit in practice, but not impossible I don't think) it could have posted a valid frame at this point.

Reply:
Good catch, fixed in 96764b578.